File: lidar_msgs/src/lidar_obstacle_avoidance.cpp
/*
roslaunch wpr_simulation wpb_simple.launch
rosrun lidar_msgs lidar_obstacle_avoidance
*/

#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <geometry_msgs/Twist.h>

ros::Publisher vel_pub;
int nCount = 0;

void lidar_callback(sensor_msgs::LaserScan msg)
{
    float distanceToForward = msg.ranges[180];
    ROS_INFO_STREAM("distance to forward:" << std::to_string(distanceToForward) << "\n");

    geometry_msgs::Twist cmd_vel;
    if(distanceToForward < 1.5 || (nCount > 0 && nCount < 50))
    {
        cmd_vel.angular.z = 0.3;
        nCount+=1;
        if(nCount == 50)
        {
            nCount = 0;
        }
    }
    else
    {
        cmd_vel.linear.x = 0.1;
    }
    vel_pub.publish(cmd_vel);
}

int main(int argc, char *argv[])
{
    ros::init(argc, argv, "lidar_msg");

    ros::NodeHandle nh;

    ros::Subscriber subscriber = nh.subscribe<sensor_msgs::LaserScan>("/scan", 3, lidar_callback);
    vel_pub = nh.advertise<geometry_msgs::Twist>("/cmd_vel", 3);

    ros::Rate rate = ros::Rate(10);

    while(ros::ok())
    {
        ros::spinOnce();
        rate.sleep();
    }
    /* code */
    return 0;
}

File: lidar_msgs/scripts/lidar_obstacle_avoidance.py
#!/usr/bin/env python3

"""
chmod +x scripts/lidar_msgs.py
roslaunch wpr_simulation wpb_simple.launch
rosrun lidar_msgs lidar_obstacle_avoidance.py
"""
import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist

nCount = 0

def lidar_callback(msg: LaserScan):
    global nCount
    
    distanceToForward = msg.ranges[180]
    rospy.loginfo(f'DistanceToForward: {distanceToForward}')

    cmd_vel = Twist()
    if distanceToForward < 1.5 or (nCount> 0 and nCount<50):
        cmd_vel.angular.z = 0.5
        
        nCount+=1
        if nCount==50:
            nCount = 0
    else:
        cmd_vel.linear.x = 0.1
    
    global vel_pub
    vel_pub.publish(cmd_vel)
    

if __name__ == "__main__":
    rospy.init_node("lidar_msgs")
    rospy.Subscriber("/scan", LaserScan, queue_size=3, callback=lidar_callback)
    
    vel_pub = rospy.Publisher("/cmd_vel", Twist, queue_size=3)
    rospy.spin()